lib/pc_patch.c

```diff
@@ -659,3 +659,70 @@ pc_patch_transform(const PCPATCH *patch, const PCSCHEMA *new_schema)
 
     return (PCPATCH*) paout;
 }
+
+/** rotate a patch given a unit quaternion */
+PCPATCH *
+pc_patch_rotate_quaternion(
+    const PCPATCH *patch,
+    double qw, double qx, double qy, double qz,
+    const char *dimx, const char *dimy, const char *dimz)
+{
+    PCPATCH_UNCOMPRESSED *paout;
+    const PCSCHEMA *schema;
+    PCPOINTLIST *opl, *npl;
+    const PCMAT44 *qmat;
+    PCVEC4 *vec, *rotatedvec;
+    size_t i, j;
+
+    vec = pcalloc(sizeof(PCVEC4));
+    if ( NULL == vec )
+        return NULL;
+
+    rotatedvec = pcalloc(sizeof(PCVEC4));
+    if ( NULL == rotatedvec )
+        return NULL;
+
+    qmat = pc_matrix_create_from_quaternion(qw, qx, qy, qz);
+    if ( NULL == qmat )
+        return NULL;
+
+    paout = pc_patch_uncompressed_make(patch->schema, patch->npoints);
+    if ( NULL == paout )
+        return NULL;
+
+    schema = patch->schema;
+
+    opl = pc_pointlist_from_patch(patch);
```

Comment:
I think the pointlist is a temporary copy already so that it can be modified in place, but that should be checked obviously.